API.md

```diff
@@ -0,0 +1,319 @@
+
+An **OpenSCD menu plugin** is an OpenSCD plugin with an additional `run()` instance method which is called when the plugin's entry is clicked in OpenSCD's menu. It is continuously rendered into the app and is expected to normally not display any content of its own. It is meant for one-shot editing tasks or tasks that always run in the background (like validation).
+
+An **OpenSCD editor plugin** is a modal OpenSCD plugin that is only rendered as long as the user has its tab selected in OpenSCD's tab bar. It is meant for rendering the main part of OpenSCD's user interface.
+
+The **OpenSCD core API** is:
+- the way in which OpenSCD core communicates relevant data to the plugins and
+- the way in which plugins communicate user intent to OpenSCD core.
+- the way in which OpenSCD sets CSS fonts and colors for plugins
+
+## Communicating data to plugins
+
+OpenSCD core communicates the data necessary for editing SCL documents by setting the following [properties](https://developer.mozilla.org/en-US/docs/Glossary/Property/JavaScript) on the plugin's [DOM Element](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement):
+
+
+```typescript=
+export default class Plugin extends LitElement {
+  docs: Record<string, XMLDocument> = {};
+  doc?: XMLDocument;
+  docName?: string;
+  editable: string[] = [
+    'cid',
+    'icd',
+    'iid',
+    'scd',
+    'sed',
+    'ssd',
+  ];
+  history: LogEntry[];
+  editCount: number = 0;
+  plugins: { menu: Plugin[], editor: Plugin[] }[];
+  locale: string = 'en';
+}
+
+/** Helper types exported by OpenSCD core **/
+
+type LogEntry = {
+  redo: Edit; // `Edit` defined below
+  undo: Edit;
+  title?: string
+}
+
+type Plugin = {
+  name: string;
+  translations?: Record<string, string>;
+  src: string;
+  icon: string;
+  requireDoc?: boolean; // disable plugin if no doc is opened
+}
+```
+
+### `docs`
+
+`docs` is set to an object mapping `string` keys (document names) to `XMLDocument` values.
+
+### `docName`
+The name of the `XMLDocument` currently being edited.
+
+### `doc`
+The `XMLDocument` currently being edited.
+
+### `editable`
+Filename extensions of user-editable documents.
+
+### `history`
+History of edits done to `doc`.
+
+### `editCount`
+
+Index of the current log entry in the `history`. Is incremented with every new edit and updated on undo/redo. 
+
+### `plugins`
+
+Arrays of `Plugin` objects describing the currently loaded `menu` and `editor` plugins, respectively.
+
+### `locale`
+
+Selected language (IETF language tag).
+
+## Communicating user intent to OpenSCD core
+
+Plugins communicate user intent to OpenSCD core by dispatching the following [custom events](https://developer.mozilla.org/en-US/docs/Web/API/CustomEvent):
+
+### `EditEventV2`
+
+The **edit event** allows a plugin to describe the changes it wants to make to the current `doc`.
+
+```typescript=
+export type EditDetailV2<E extends Edit = Edit> = {
+  edit: E;
+  title?: string;
+  squash?: boolean;
```

Comment:
On contemplation this does seem like an elegant way to resolve the poorly worded concerns ([here](https://github.com/OMICRONEnergy/oscd-designer/issues/57)) without introducing new problems.

What does "squash" do? Is it purely a "grouping after the event" (literally).
What if a plugin (e.g. a menu plugin) listens to another plugin and pushes out events in response to an EditEvent. 

Would this result in these edits being squashed or handled in an unexpected fashion? I'm particularly thinking of what I do in the subscriber plugin. I register a `window` event ([here](https://github.com/danyill/oscd-subscriber-later-binding/blob/670598f6cc5c64168f3e9b59eef8c7e03eadf886/oscd-subscriber-later-binding.ts#L546-L557 )) and then the IED specific plugins generate new events if appropriate ([here](https://github.com/danyill/oscd-subscriber-lb-siemens/blob/ae20db2895c4c9f73029d090d3fae0019ae51854/oscd-subscriber-lb-siemens.ts#L217-L227 ), it may not be done well!). I guess in this case I would like the dependent events to be included in the initiating event but perhaps this wouldn't always be the case?

Reply:
Whenever a PR is coming with the `squash` flag, the editing handle does merge it with the edit before. This would allow a plugin author to avoid unwanted redo states. For your example, you could use the flag to sqash the edit coming from the additional menu plugins with the ones that triggered the menu plugin. 

But I agree in general. The event handling in the browser does allow you to listen to any events and send your own events. We therefore cannot guarantee the there is not an in between edit event coming in and spoling the sqash.

Reply:
Thinking about it I feel that plugins and dependent plugins can sort out their own in-house rules and if they spoil the squash, it's on them. They can use the composed path to access information from the event source like I do now:

```js
function shouldListen(event: Event): boolean {
  const initiatingTarget = <Element>event.composedPath()[0];
  return (
    initiatingTarget instanceof Element &&
    initiatingTarget.getAttribute('identity') ===
      'danyill.oscd-subscriber-later-binding' &&
    initiatingTarget.hasAttribute('allowexternalplugins')
  );
}
```

So I think no change or improvement is required.

Reply:
Thank you both for the discussion. My takeaway is that the behaviour of the `squash` flag and its use cases need to be made very clear in the eventual API documentation (generated from code).